### DISCUSSION TITLE
How to let image texture take 3d object coordinate as input?

### DISCUSSION THREAD

**Discussion by HSDSZ** [bnpr/Malt#546] [Q&A] [answered]
I'm planning to make a procedual landscape material. I am using object coordinate instead uv coordinate for image texture (landscape is also procedual, so we don't have uv). 
However, malt and eevee treat texture coordinate differently
here is eevee
![image](https://github.com/bnpr/Malt/assets/36077492/4ecb57b6-623d-4290-8f21-ff7acc2ecd9b)
in malt
since malt texture only take 2d vector as scale, even i use object coordinate like this [#159](https://github.com/bnpr/Malt/discussions/159)

the result looks weird because one direction of object coordinate is ignored. Like below, z axis is ignored, so texture in z is scaled too much
![image](https://github.com/bnpr/Malt/assets/36077492/f36d5a0b-74d7-4006-bd7e-ac05a7600cbe)

**Comment by HSDSZ** [comment]
[objectcoord.zip](https://github.com/bnpr/Malt/files/13932532/objectcoord.zip)

**Comment by DeltishEffort** [comment, marked as answer]
So I believe the EVEE image texture node can receive a vec3 UV and handles the projection method which also requires the generated Normals to decide how the object coordinates being used in place of the UVs will map to the image. 

I am not sure if there is a built in way to do this right now but luckily Malt is amazing and flexible!

I came across a video a little while ago meant for godot about cheap non triplanar mapping, it was helpful for me to get a better idea about some of this. here is a Malt node version of essentially that. But the downside is no fancy blending like in vanilla Evee.
![image](https://github.com/bnpr/Malt/assets/41149511/72be4d94-fdca-4b7e-871e-1e5daa00c1fe)

But we also can borrow some of the shader code from Blender's source and implement it as a Malt plugin perhaps?
Here is an example of what that would look like.
![image](https://github.com/bnpr/Malt/assets/41149511/f9558b26-c40a-4e72-8cd7-a9a1c67e0bcd)
And here is the [repo for the plugin example](https://github.com/DeltishEffort/ImagePlusMaltPlugin/tree/main)

Some comparisons
![Nodes_no_blend](https://github.com/bnpr/Malt/assets/41149511/5e7e7bd4-0b93-47fd-96f0-00bff8a6cc70)
![malt_plugin_node](https://github.com/bnpr/Malt/assets/41149511/c2f5e844-4d4e-4b36-baed-946efe21e402)
![evee_result](https://github.com/bnpr/Malt/assets/41149511/47681822-be06-468b-b1ad-944fd6f8e851)

Hope this helps
INB4 someone comes along and says there is a simpler way lol

**Comment by Dereuz24** [comment]
Hi @HSDSZ Sorry to revive this old thread but II found this post trying to figure out how to create UVless decals with a similar method and needed some help

I was thinking on maybe using just a simple empty perpendicular to the surface and take the empty position as an input and project the image in a similar manner but based on the empty position and scale just like one would do on eevee. 

But for that to work what else should I change? I was thinking on using a second geometry input node for the empty but also I was thinking on how a lot of decal nodes on eevee have a treshold to control how much of the image is projected on the axis of the empty to account for flat tapers and to flood objects that are behind the main object

The geometry node input is self referential so, would I have to use the inline code node for getting the position of another object? What if I wanted to update the object used every time I want to create a decal? Do I have to write a node to avoid having to repeat that process every time? I really don't know much about this so I just wanted to be pointed in the right direction
